add kube helper

farshad68 opened the pull request.

MichaelRoeder committed:
Increased the version to 1.0.24-SNAPSHOT.

farshad68 committed:
add kube helper

MichaelRoeder commented on src/main/java/org/hobbit/utils/kubernetes/KubeHelper.java:
```diff
@@ -0,0 +1,40 @@
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+
+public class KubeHelper {
```
`DNSHelper`

farshad68 replied:
Also, I add DNSHelper asa  class to use by modules and that class will use this Kubernetes class, then I keep the name

MichaelRoeder commented on src/main/java/org/hobbit/utils/kubernetes/KubeHelper.java:
```diff
@@ -0,0 +1,40 @@
+package org.hobbit.utils.kubernetes;
```
`org.hobbit.utils` is sufficient

MichaelRoeder commented on src/main/java/org/hobbit/utils/kubernetes/KubeHelper.java:
```diff
@@ -0,0 +1,40 @@
+public class KubeHelper {
+    private static final Logger LOGGER = LoggerFactory.getLogger(KubeHelper.class);
+    /**
+     * Gets dns friendly ip
```
"Get network identifier that can be used to reach this container. This can be either a DNS name, an IP address or something similar."

farshad68 replied:
here for Kubernetes it would be IP

MichaelRoeder commented on src/main/java/org/hobbit/utils/kubernetes/KubeHelper.java:
```diff
@@ -0,0 +1,40 @@
+     *
+     * @return plain hostname (or address)
+     */
+    public static String getDnsFriendlyIP(String namespace) {
```
`getDNSName`

farshad68 replied:
I think the method name defines the functionality of the method better, so maybe is better to keep it

MichaelRoeder commented on src/main/java/org/hobbit/utils/kubernetes/KubeHelper.java:
```diff
@@ -0,0 +1,40 @@
+        }
+    }
+
+    public static String getPodIP() {
```
`getContainerIp`

farshad68 replied:
because we define another class (DNSHelper ) then here we just talk about Kubernetes , then  Pod is a better match

MichaelRoeder reviewed (changes requested):
The class looks good but I would suggest to have a more "neutral" usage of terms, i.e., we are not looking for pod names and this is a class that should not only work with kubernetes :wink:

denkv commented on pom.xml:
```diff
@@ -12,7 +12,7 @@
     <groupId>org.hobbit</groupId>
     <artifactId>core</artifactId>
-    <version>1.0.23</version>
+    <version>1.0.25-SNAPSHOT</version>
```
Should we remove `SNAPSHOT` before merging into `master`?

farshad68 replied:
yes

denkv commented on src/main/java/org/hobbit/utils/kubernetes/KubeHelper.java:
```diff
@@ -0,0 +1,40 @@
+            namespace = "default";
+        }
+        String ip =  getPodIP();
+        if(!ip.equals("unknown")&&ip!=null) {
```
Would be nice to have a link in comments to a documentation or other information regarding when it is `unknown`.

denkv commented on src/main/java/org/hobbit/utils/kubernetes/KubeHelper.java:
```diff
@@ -0,0 +1,40 @@
+        } catch (UnknownHostException e) {
+            // Handle the exception, e.g., log an error or return a default value.
+            LOGGER.error("Error getting pod IP: " + e.getMessage());
+            return "unknown"; // Or throw a RuntimeException, etc.
```
Does it make sense to not fail and continue the experiment in this case? While some containers might not need themselves to be addressable, we currently don't have a way to tell this in advance.

farshad68 replied:
you are right, in this case, it is better to fail (now it will throw an exception in getDnsFriendlyIP ), but anyway the unknown is not reachable and the module will fail in the next step.

MichaelRoeder replied:
Do we really need "unknown", here? Wouldn't returning `null` make more sense? This would also fit very well to the if statement in the other method, that checks for `null` anyway...

farshad68 commented:
> The class looks good but I would suggest to have a more "neutral" usage of terms, i.e., we are not looking for pod names and this is a class that should not only work with kubernetes 😉

I see that this is a Kubernetes-specific helper, similar to how the Docker version exists in the core. Given that, I’m not sure I fully understand your point about making it work beyond Kubernetes. Right now, the client (any module using this) needs to determine whether it's running in a Kubernetes environment. If it is, it would use this helper to get the host.

Do you have a different approach in mind?

MichaelRoeder commented:
That is a good question :slightly_smiling_face: 

From my perspective, we have a program and it simply wants to get the address at which it can be reached. It calls this helper class and the class will give it the information it needs. Note the following implications:
1. The program that calls this helper class doesn't care in which environment it is running. It just wants to have the name of the node in the network. So the program shouldn't have to ask itself "Am I running on Kubernetes?".
2. This helper class should take care of it and try to detect based on the environment whether it is running in Kubernetes. Then, it should derive the String, etc.

I even would raise the question whether the "namespace" string is really necessary... shouldn't this be something that the helper class can derive, somehow?

farshad68 committed:
Add dnsHelper for module usage (backward compatible)

- Introduced dnsHelper to be used by modules.

- Ensures backward compatibility.

- No changes made to the DockerHelper class.

denkv commented on src/main/java/org/hobbit/utils/DNSHelper.java:
```diff
@@ -0,0 +1,21 @@
+    private static final Logger LOGGER = LoggerFactory.getLogger(DNSHelper.class);
+    public static String getDNSName() {
+        // determine if it is in docker or Kubernetes
+        String isKubernetes = System.getenv("IS_KUBERNETES");
```
What if we get more options in the future? Maybe a variable with a value indicating which exactly it is (docker swarm, kubernetes or something else)?

farshad68 replied:
then we can call the variable RUN_ON and the value could be "KUBERNETES" or "DOCKER" is it good or do you have another suggestion ?

farshad68 committed:
throw exception if can not get DNS friendly IP for any reason

denkv commented on src/main/java/org/hobbit/utils/DNSHelper.java:
```diff
@@ -0,0 +1,21 @@
+    public static String getDNSName() {
+        // determine if it is in docker or Kubernetes
+        String isKubernetes = System.getenv("IS_KUBERNETES");
+        if(isKubernetes != null && isKubernetes.equalsIgnoreCase("true")) {
```
https://github.com/hobbit-project/core/blob/aa0ea3852dd3ff25c29ff7f5fae2b54b60f3b6b7/src/main/java/org/hobbit/utils/EnvVariables.java#L817

farshad68 committed:
refactor code , and use RUN_ON as environment variable to decide it is kubernetes or docker or anything in future :D

farshad68 committed:
add test

denkv commented on src/main/java/org/hobbit/utils/kubernetes/KubeHelper.java:
```diff
@@ -0,0 +1,72 @@
+     * @return The value of the environment variable or an empty string if not found.
+     */
+    protected static String getEnvVariable(String key) {
+        return Optional.ofNullable(System.getenv(key)).orElse("").trim();
```
Why not use this helper from this library directly?

https://github.com/hobbit-project/core/blob/aa0ea3852dd3ff25c29ff7f5fae2b54b60f3b6b7/src/main/java/org/hobbit/utils/EnvVariables.java#L117

farshad68 replied:
thanks for mentioning it, I was not aware of it :)

farshad68 replied:
but it is "Deprecated" :/
I will use org.hobbit.utils.config.HobbitConfiguration

farshad68 committed:
change codacy reporter source

farshad68 committed:
change codacy setup

farshad68 committed:
use HobbitConfiguration for getting the environment variables

farshad68 committed:
return null when can not get IP

MichaelRoeder reviewed (commented):
LGTM
